OpInfo for `*_like` functions

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66358 OpInfos for: normal, bernoulli, multinomial
* **#65941 OpInfo for `*_like` functions**
* #64282 OpInfos for some Tensor dtype conversion methods

OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

zou3519 committed:
OpInfo for `*_like` functions

OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 commented:
NB: I will fix the lint whenever fb comes back up...

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2015,6 +2015,27 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
 
     return tuple(samples)
 
+def sample_inputs_like_fns(self, device, dtype, requires_grad, full_like=False):
```
All sample inputs functions should take `**kwargs`

```suggestion
def sample_inputs_like_fns(self, device, dtype, requires_grad, full_like=False, **kwargs):
```

mruberry replied:
Some sample inputs take optional kwargs that tests can opportunistically set:

https://github.com/pytorch/pytorch/blob/6d4d636d66774eb803cffedd267849a8b8a3b536/torch/testing/_internal/common_methods_invocations.py#L4353

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9327,6 +9348,57 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
+           sample_inputs_func=sample_inputs_like_fns,
+           supports_autograd=False,
+           skips=(
+               # This method not supported by TorchScript
+               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
```
See https://github.com/pytorch/pytorch/pull/64282#discussion_r722917935.

```suggestion
               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit', dtypes=(torch.float32, torch.complex64)),
```

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2015,6 +2015,27 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
+        t = make_tensor(shape, device, dtype,
+                        low=None, high=None,
+                        requires_grad=requires_grad)
+        if full_like:
+            samples.append(SampleInput(t, args=(123,), kwargs=kwargs))
+        else:
+            samples.append(SampleInput(t, kwargs=kwargs))
```
I dislike the special casing here for a single op. I would prefer to have a separate sample inputs function for it instead:

```python
def sample_inputs_full_like(*args, fill_value=123, **kwargs):
    sample_inputs = []
    for sample_input in sample_inputs_like_fns(*args, **kwargs):
        sample_input.args = (fill_value,)
        sample_inputs.append(sample_input)
    return sample_inputs
```

pmeier reviewed (commented):
Some comments inline. Otherwise LGTM when CI is green.

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2015,6 +2015,27 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
+        ((), {}),
+        ((S, S), {}),
+        ((0, S, 0), {}),
+        ((S,), {'dtype': torch.double}),
+        ((S,), {'device': 'cpu'}),
+        ((S,), {'dtype': torch.double, 'device': 'cpu'}),
```
I wanted to test some cases where the (dtype, device) being passed to e.g. empty_like is different from the input Tensor's (dtype, device)

mruberry replied:
It'd be cool to conditionally add CUDA too (the function can check if CUDA is available).

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2021,6 +2021,35 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
     return tuple(samples)
 
+def sample_inputs_like_fns(self, device, dtype, requires_grad, **kwargs):
+    inputs: List[Tuple[Tuple, Dict]] = [
```
Meta note: typing is lame -- did something complain this wasn't done? If so we should fix that so typing is never required in this file.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2021,6 +2021,35 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
+
+    return tuple(samples)
+
+def sample_inputs_full_like(self, *args, **kwargs):
```
Cool

mruberry reviewed (approved):
Cool!

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

[ghstack-poisoned]

pmeier reviewed (approved):
I second https://github.com/pytorch/pytorch/pull/65941#discussion_r723576574, otherwise LGTM!

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

zou3519 committed:
Update on "OpInfo for `*_like` functions"


OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).
